src/Dependencies/Collections/ImmutableSegmentedList`1.cs

```diff
@@ -0,0 +1,632 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.Collections.Internal;
+
+namespace Microsoft.CodeAnalysis.Collections
+{
+    internal readonly partial struct ImmutableSegmentedList<T> : IImmutableList<T>, IReadOnlyList<T>, IList<T>, IList, IEquatable<ImmutableSegmentedList<T>>
+    {
+        /// <inheritdoc cref="ImmutableList{T}.Empty"/>
+        public static readonly ImmutableSegmentedList<T> Empty = new(new SegmentedList<T>());
+
+        private readonly SegmentedList<T> _list;
+
+        private ImmutableSegmentedList(SegmentedList<T> list)
+            => _list = list;
+
+        public int Count => _list.Count;
+
+        /// <inheritdoc cref="ImmutableList{T}.IsEmpty"/>
+        public bool IsEmpty => _list.Count == 0;
+
+        bool ICollection<T>.IsReadOnly => true;
+
+        bool IList.IsFixedSize => true;
+
+        bool IList.IsReadOnly => true;
+
+        bool ICollection.IsSynchronized => true;
+
+        object ICollection.SyncRoot => _list;
+
+        public T this[int index] => _list[index];
+
+        T IList<T>.this[int index]
+        {
+            get => _list[index];
+            set => throw new NotSupportedException();
+        }
+
+        object? IList.this[int index]
+        {
+            get => _list[index];
+            set => throw new NotSupportedException();
+        }
+
+        public static bool operator ==(ImmutableSegmentedList<T> left, ImmutableSegmentedList<T> right)
```

Comment:
Do users expect this sort of thing to be overridden with an intelligent implementation?

Reply:
➡️ `ImmutableList<T>` is a reference type with reference equality semantics. `ImmutableSegmentedList<T>` is a value type, and `operator==` is implemented to preserve the semantics you would get for the same operator on `ImmutableList<T>`.